File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/src/domain/mutation_client.dart
/// Mutation domain client for Grey SDK.
library;

import 'dart:async';

import 'package:grpc/grpc.dart';

import '../error/grey_error.dart';
import '../grpc/grey_channel.dart';
import '../grpc/mutation_grpc_service.dart';
import '../grpc/types.dart';

/// Client for generic mutation operations.
class MutationClient {
  /// Creates a new mutation client.
  MutationClient(this._channel, this._service);

  final GreyChannel _channel;
  final MutationGrpcService _service;

  /// Executes a mutation against the specified endpoint.
  ///
  /// [endpoint] is the mutation endpoint path.
  /// [method] is the HTTP method equivalent (default: 'POST').
  /// [body] is the optional request body.
  ///
  /// Throws [GreyError] if not authenticated or on failure.
  Future<MutationData> mutate({
    required String endpoint,
    String method = 'POST',
    Object? body,
  }) async {
    // Validate inputs
    if (endpoint.isEmpty) {
      throw GreyError(
        code: GreyError.validationError,
        message: 'Endpoint is required',
      );
    }

    // Validate method
    const validMethods = ['POST', 'PUT', 'PATCH', 'DELETE'];
    if (!validMethods.contains(method.toUpperCase())) {
      throw GreyError(
        code: GreyError.validationError,
        message: 'Method must be one of: ${validMethods.join(', ')}',
      );
    }

    // Check authentication
    if (!_channel.isAuthenticated) {
      throw GreyError(
        code: GreyError.unauthorized,
        message: 'Not authenticated',
      );
    }

    try {
      return await _service.mutate(
        endpoint: endpoint,
        method: method.toUpperCase(),
        body: body,
      );
    } on GrpcError catch (e) {
      throw GreyError.fromGrpcError(e);
    } catch (e) {
      throw GreyError.fromException(e);
    }
  }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-scala/src/main/scala/com/grey/sdk/domain/MutationClient.scala
package com.grey.sdk.domain

import com.grey.sdk.error.GreyError
import com.grey.sdk.grpc.{GreyChannel, MutationData, MutationGrpcService}
import scala.concurrent.{ExecutionContext, Future}

/** Client for generic mutation operations.
  *
  * Provides a flexible mutation interface for custom endpoints.
  */
final class MutationClient(
    channel: GreyChannel,
    service: MutationGrpcService
)(using ec: ExecutionContext):

  private val ValidMethods: Set[String] = Set("POST", "PUT", "PATCH", "DELETE")

  /** Executes a mutation against the specified endpoint.
    *
    * @param endpoint
    *   The mutation endpoint path
    * @param method
    *   The HTTP method equivalent (default: "POST")
    * @param body
    *   Optional request body
    * @return
    *   Future containing MutationData on success
    * @throws GreyError
    *   if not authenticated or on failure
    */
  def mutate(
      endpoint: String,
      method: String = "POST",
      body: Option[Any] = None
  ): Future[MutationData] =
    // Validate inputs
    if endpoint.isEmpty then
      return Future.failed(GreyError.validation("Endpoint is required"))

    val normalizedMethod = method.toUpperCase
    if !ValidMethods.contains(normalizedMethod) then
      return Future.failed(
        GreyError.validation(s"Method must be one of: ${ValidMethods.mkString(", ")}")
      )

    // Check authentication
    if !channel.isAuthenticated then
      return Future.failed(GreyError.unauthorized())

    service.mutate(endpoint, normalizedMethod, body)
      .recoverWith { case e => Future.failed(GreyError.fromThrowable(e)) }

  /** Executes mutation and returns Either instead of throwing. */
  def mutateSafe(
      endpoint: String,
      method: String = "POST",
      body: Option[Any] = None
  ): Future[Either[GreyError, MutationData]] =
    mutate(endpoint, method, body)
      .map(Right(_))
      .recover { case e: GreyError => Left(e) }

object MutationClient:
  def apply(channel: GreyChannel, service: MutationGrpcService)(using ExecutionContext): MutationClient =
    new MutationClient(channel, service)
